The Alevin and Alevin-fry (latest improvement) algorithms claims to perform better than CellRanger and we tried it on p.falciparum data.

# Alevin












## Resources Referred
1. [Alevin Manual](https://salmon.readthedocs.io/en/latest/alevin.html#alevin-logs)
2. [alevinQC](https://csoneson.github.io/alevinQC/)
3. [Selective Alignment](https://combine-lab.github.io/alevin-tutorial/2019/selective-alignment/)
4. [Run Salmon](https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/04_quasi_alignment_salmon.html)
5. [How does salmon deal with decoy?](https://www.biostars.org/p/456231/)
6. [Setting Up Resources](https://combine-lab.github.io/alevin-tutorial/2018/setting-up-resources/)
7. [alevin-default.ipynb](https://gist.github.com/k3yavi/c501705ed2d29b12b0d10cf78b3ed001)
8. [STARsolo and Alevin-full-decoy: High Speed and High Accuracy](https://www.singlecellcourse.org/processing-raw-scrna-seq-sequencing-data-from-reads-to-a-count-matrix.html#starsolo-and-alevin-full-decoy-high-speed-and-high-accuracy)
9. [Single cell tutorial (using UMI tools and STAR)](https://umi-tools.readthedocs.io/en/latest/Single_cell_tutorial.html)
10. [Alevin efficiently estimates accurate gene abundances from dscRNA-seq data](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1670-y) 
11. [Generating a single cell matrix using Alevin](https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/droplet-quantification-preprocessing/tutorial.html#generate-a-transcriptome-index--quantify)
12. [GITTER Alevin discussion](https://gitter.im/COMBINE-lab/salmon?at=5dcea2c1ca778c1fbf09a6b0)
13. [How to change cellranger multimapping algorithm](https://www.biostars.org/p/356854/)
14. [Running the alevin-fry pipeline](https://alevin-fry.readthedocs.io/en/latest/getting_started.html)
15. [Alevin-fry tutorial](https://alevin-fry.readthedocs.io/en/latest/getting_started.html)
